File: ui-swift/SecretKeeper/Sources/SecretKeeper/Services/IPCClient.swift
import Foundation

protocol IPCClientDelegate: AnyObject {
    func ipcClient(_ client: IPCClient, didReceiveViolation violation: ViolationEvent)
    func ipcClient(_ client: IPCClient, didReceiveViolationHistory violations: [ViolationEvent])
    func ipcClient(_ client: IPCClient, didUpdateStatus status: AgentStatus)
    func ipcClient(_ client: IPCClient, didReceiveCategories categories: [ProtectedCategory])
    func ipcClient(_ client: IPCClient, didReceiveAgentInfo info: AgentInfo)
    func ipcClientDidConnect(_ client: IPCClient)
    func ipcClientDidDisconnect(_ client: IPCClient)
}

/// Agent binary info for auto-upgrade detection.
struct AgentInfo {
    /// Binary modification time (Unix timestamp in seconds).
    let binaryMtime: Int64
    /// Agent version string.
    let version: String
}

/// Client for communicating with the SecretKeeper agent via Unix socket.
class IPCClient: NSObject {
    private let socketPath: String
    private var inputStream: InputStream?
    private var outputStream: OutputStream?
    private var isConnected = false
    private let queue = DispatchQueue(label: "com.codegroove.secretkeeper.ipc", qos: .userInitiated)
    private var readBuffer = Data()

    weak var delegate: IPCClientDelegate?

    private let decoder: JSONDecoder = {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return decoder
    }()

    private let encoder: JSONEncoder = {
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        return encoder
    }()

    init(socketPath: String) {
        self.socketPath = socketPath
    }

    /// Connect to the agent socket.
    func connect() {
        queue.async { [weak self] in
            self?.doConnect()
        }
    }

    private func doConnect() {
        guard !isConnected else { return }

        // Create Unix domain socket streams
        var readStream: Unmanaged<CFReadStream>?
        var writeStream: Unmanaged<CFWriteStream>?

        CFStreamCreatePairWithSocketToHost(
            kCFAllocatorDefault,
            nil,
            0,
            &readStream,
            &writeStream
        )

        // For Unix sockets, we need to use a different approach
        let socketHandle = socket(AF_UNIX, SOCK_STREAM, 0)
        guard socketHandle >= 0 else {
            print("Failed to create socket")
            return
        }

        var addr = sockaddr_un()
        addr.sun_family = sa_family_t(AF_UNIX)

        // Copy socket path safely with bounds checking
        let maxPathLen = MemoryLayout.size(ofValue: addr.sun_path) - 1  // Reserve space for null terminator
        guard socketPath.utf8.count <= maxPathLen else {
            print("Socket path too long: \(socketPath.utf8.count) > \(maxPathLen)")
            close(socketHandle)
            return
        }

        socketPath.withCString { path in
            withUnsafeMutablePointer(to: &addr.sun_path) { sunPath in
                let ptr = UnsafeMutableRawPointer(sunPath).assumingMemoryBound(to: CChar.self)
                strncpy(ptr, path, maxPathLen)
                ptr[maxPathLen] = 0  // Ensure null termination
            }
        }

        let connectResult = withUnsafePointer(to: &addr) { addrPtr in
            addrPtr.withMemoryRebound(to: sockaddr.self, capacity: 1) { sockaddrPtr in
                Darwin.connect(socketHandle, sockaddrPtr, socklen_t(MemoryLayout<sockaddr_un>.size))
            }
        }

        guard connectResult == 0 else {
            print("Failed to connect to \(socketPath): \(String(cString: strerror(errno)))")
            close(socketHandle)
            return
        }

        // Create streams from socket
        CFStreamCreatePairWithSocket(
            kCFAllocatorDefault,
            Int32(socketHandle),
            &readStream,
            &writeStream
        )

        guard let input = readStream?.takeRetainedValue(),
              let output = writeStream?.takeRetainedValue() else {
            print("Failed to create streams")
            close(socketHandle)
            return
        }

        inputStream = input as InputStream
        outputStream = output as OutputStream

        CFReadStreamSetProperty(input, CFStreamPropertyKey(rawValue: kCFStreamPropertyShouldCloseNativeSocket), kCFBooleanTrue)
        CFWriteStreamSetProperty(output, CFStreamPropertyKey(rawValue: kCFStreamPropertyShouldCloseNativeSocket), kCFBooleanTrue)

        inputStream?.delegate = self
        outputStream?.delegate = self

        inputStream?.schedule(in: .main, forMode: .common)
        outputStream?.schedule(in: .main, forMode: .common)

        inputStream?.open()
        outputStream?.open()

        isConnected = true

        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            self.delegate?.ipcClientDidConnect(self)
        }

        // Start reading
        startReading()
    }

    /// Disconnect from the agent.
    func disconnect() {
        queue.async { [weak self] in
            self?.doDisconnect()
        }
    }

    private func doDisconnect() {
        inputStream?.close()
        outputStream?.close()
        inputStream = nil
        outputStream = nil
        isConnected = false

        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            self.delegate?.ipcClientDidDisconnect(self)
        }
    }

    /// Subscribe to real-time events.
    func subscribe(filter: EventFilter? = nil) {
        let request = SubscribeRequest(filter: filter)
        send(request)
    }

    /// Get agent status.
    func getStatus() {
        send(StatusRequest())
    }

    /// Get current mode.
    func getMode() {
        send(GetModeRequest())
    }

    /// Set enforcement mode.
    func setMode(_ mode: EnforcementMode) {
        send(SetModeRequest(mode: mode.rawValue))
    }

    /// Allow a suspended process to continue (one-time).
    func allowOnce(eventId: String) {
        send(AllowOnceRequest(eventId: eventId))
    }

    /// Add permanent/temporary exception for a violation.
    func allowPermanently(eventId: String, expiresAt: Date? = nil, comment: String? = nil) {
        send(AllowPermanentlyRequest(eventId: eventId, expiresAt: expiresAt, comment: comment))
    }

    /// Kill a suspended process.
    func killProcess(eventId: String) {
        send(KillRequest(eventId: eventId))
    }

    /// Get violation history.
    func getViolations(limit: Int? = nil, since: Date? = nil) {
        send(GetViolationsRequest(limit: limit, since: since))
    }

    /// Get all active exceptions.
    func getExceptions() {
        send(GetExceptionsRequest())
    }

    /// Add a new exception.
    func addException(
        processPath: String? = nil,
        codeSigner: String? = nil,
        filePattern: String,
        isGlob: Bool = true,
        expiresAt: Date? = nil,
        comment: String? = nil
    ) {
        send(AddExceptionRequest(
            processPath: processPath,
            codeSigner: codeSigner,
            filePattern: filePattern,
            isGlob: isGlob,
            expiresAt: expiresAt,
            comment: comment
        ))
    }

    /// Remove an exception by ID.
    func removeException(id: Int64) {
        send(RemoveExceptionRequest(id: id))
    }

    /// Ping the agent.
    func ping() {
        send(PingRequest())
    }

    /// Get all protected categories with their enabled status.
    func getCategories() {
        send(GetCategoriesRequest())
    }

    /// Enable or disable a protected category.
    func setCategoryEnabled(categoryId: String, enabled: Bool) {
        send(SetCategoryEnabledRequest(categoryId: categoryId, enabled: enabled))
    }

    /// Get agent binary info for auto-upgrade detection.
    func getAgentInfo() {
        send(GetAgentInfoRequest())
    }

    // MARK: - Private

    private func send<T: Encodable>(_ request: T) {
        queue.async { [weak self] in
            self?.doSend(request)
        }
    }

    private func doSend<T: Encodable>(_ request: T) {
        guard isConnected, let output = outputStream else {
            print("Not connected")
            return
        }

        do {
            var data = try encoder.encode(request)
            data.append(contentsOf: [0x0A]) // Newline

            let bytes = [UInt8](data)
            let written = output.write(bytes, maxLength: bytes.count)

            if written < 0 {
                print("Write error: \(output.streamError?.localizedDescription ?? "unknown")")
            }
        } catch {
            print("Encode error: \(error)")
        }
    }

    private func startReading() {
        queue.async { [weak self] in
            self?.readLoop()
        }
    }

    private func readLoop() {
        guard isConnected, let input = inputStream else { return }

        var buffer = [UInt8](repeating: 0, count: 4096)

        while input.hasBytesAvailable {
            let bytesRead = input.read(&buffer, maxLength: buffer.count)

            if bytesRead > 0 {
                readBuffer.append(contentsOf: buffer[0..<bytesRead])
                processBuffer()
            } else if bytesRead < 0 {
                print("Read error: \(input.streamError?.localizedDescription ?? "unknown")")
                doDisconnect()
                return
            }
        }

        // Continue reading
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) { [weak self] in
            self?.queue.async {
                self?.readLoop()
            }
        }
    }

    private func processBuffer() {
        // Look for newline-delimited JSON messages
        while let newlineIndex = readBuffer.firstIndex(of: 0x0A) {
            let lineData = readBuffer[..<newlineIndex]
            readBuffer = Data(readBuffer[(newlineIndex + 1)...])

            if lineData.isEmpty { continue }

            handleMessage(Data(lineData))
        }
    }

    private func handleMessage(_ data: Data) {
        // Try to determine message type
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            print("Failed to parse JSON")
            return
        }

        if let status = json["status"] as? String {
            switch status {
            case "event":
                // Violation event - ViolationEvent fields are flattened directly in the JSON
                if let eventData = try? JSONSerialization.data(withJSONObject: json) {
                    do {
                        let violation = try decoder.decode(ViolationEvent.self, from: eventData)
                        DispatchQueue.main.async { [weak self] in
                            guard let self = self else { return }
                            self.delegate?.ipcClient(self, didReceiveViolation: violation)
                        }
                    } catch {
                        fputs("[IPCClient] Failed to decode violation event: \(error)\n", stderr)
                        fputs("[IPCClient] Raw JSON: \(String(data: eventData, encoding: .utf8) ?? "nil")\n", stderr)
                    }
                }

            case "status":
                // Agent status
                if let statusData = try? JSONSerialization.data(withJSONObject: json),
                   let agentStatus = try? decoder.decode(AgentStatus.self, from: statusData) {
                    DispatchQueue.main.async { [weak self] in
                        guard let self = self else { return }
                        self.delegate?.ipcClient(self, didUpdateStatus: agentStatus)
                    }
                }

            case "categories":
                // Categories response
                if let categoriesArray = json["categories"] as? [[String: Any]] {
                    var categories: [ProtectedCategory] = []
                    for catDict in categoriesArray {
                        if let id = catDict["id"] as? String,
                           let enabled = catDict["enabled"] as? Bool,
                           let patterns = catDict["patterns"] as? [String] {
                            categories.append(ProtectedCategory(id: id, enabled: enabled, patterns: patterns))
                        }
                    }
                    DispatchQueue.main.async { [weak self] in
                        guard let self = self else { return }
                        self.delegate?.ipcClient(self, didReceiveCategories: categories)
                    }
                }

            case "agent_info":
                // Agent info response for auto-upgrade
                if let binaryMtime = json["binary_mtime"] as? Int64,
                   let version = json["version"] as? String {
                    let info = AgentInfo(binaryMtime: binaryMtime, version: version)
                    DispatchQueue.main.async { [weak self] in
                        guard let self = self else { return }
                        self.delegate?.ipcClient(self, didReceiveAgentInfo: info)
                    }
                }

            case "violations":
                // Violation history response
                if let eventsArray = json["events"] as? [[String: Any]] {
                    var violations: [ViolationEvent] = []
                    for eventDict in eventsArray {
                        if let eventData = try? JSONSerialization.data(withJSONObject: eventDict),
                           let violation = try? decoder.decode(ViolationEvent.self, from: eventData) {
                            violations.append(violation)
                        }
                    }
                    DispatchQueue.main.async { [weak self] in
                        guard let self = self else { return }
                        self.delegate?.ipcClient(self, didReceiveViolationHistory: violations)
                    }
                }

            case "success", "error", "pong", "exceptions", "config":
                // Handle other response types as needed
                break

            default:
                break
            }
        }
    }
}

extension IPCClient: StreamDelegate {
    func stream(_ aStream: Stream, handle eventCode: Stream.Event) {
        switch eventCode {
        case .errorOccurred:
            print("Stream error: \(aStream.streamError?.localizedDescription ?? "unknown")")
            doDisconnect()
        case .endEncountered:
            print("Stream ended")
            doDisconnect()
        default:
            break
        }
    }
}

// MARK: - Request Types

private struct SubscribeRequest: Codable {
    let action = "subscribe"
    let filter: EventFilter?
}

struct EventFilter: Codable {
    let filePatterns: [String]?
    let ruleIds: [String]?
    let deniedOnly: Bool?

    enum CodingKeys: String, CodingKey {
        case filePatterns = "file_patterns"
        case ruleIds = "rule_ids"
        case deniedOnly = "denied_only"
    }
}

private struct StatusRequest: Codable {
    let action = "status"
}

private struct GetModeRequest: Codable {
    let action = "get_mode"
}

private struct SetModeRequest: Codable {
    let action = "set_mode"
    let mode: String
}

private struct AllowOnceRequest: Codable {
    let action = "allow_once"
    let eventId: String

    enum CodingKeys: String, CodingKey {
        case action
        case eventId = "event_id"
    }
}

private struct AllowPermanentlyRequest: Codable {
    let action = "allow_permanently"
    let eventId: String
    let expiresAt: Date?
    let comment: String?

    enum CodingKeys: String, CodingKey {
        case action
        case eventId = "event_id"
        case expiresAt = "expires_at"
        case comment
    }
}

private struct KillRequest: Codable {
    let action = "kill"
    let eventId: String

    enum CodingKeys: String, CodingKey {
        case action
        case eventId = "event_id"
    }
}

private struct GetViolationsRequest: Codable {
    let action = "get_violations"
    let limit: Int?
    let since: Date?
}

private struct GetExceptionsRequest: Codable {
    let action = "get_exceptions"
}

private struct RemoveExceptionRequest: Codable {
    let action = "remove_exception"
    let id: Int64
}

private struct PingRequest: Codable {
    let action = "ping"
}

private struct GetCategoriesRequest: Codable {
    let action = "get_categories"
}

private struct SetCategoryEnabledRequest: Codable {
    let action = "set_category_enabled"
    let categoryId: String
    let enabled: Bool

    enum CodingKeys: String, CodingKey {
        case action
        case categoryId = "category_id"
        case enabled
    }
}

private struct GetAgentInfoRequest: Codable {
    let action = "get_agent_info"
}

File: ui-tauri/src/ipc_client.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::io;
use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
use tokio::net::UnixStream;

const SOCKET_PATH: &str = "/var/run/secretkeeper.sock";

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ViolationEvent {
    pub id: String,
    pub timestamp: DateTime<Utc>,
    pub rule_id: Option<String>,
    pub file_path: String,
    pub process_path: String,
    pub process_pid: u32,
    pub parent_pid: Option<u32>,
    pub process_euid: Option<u32>,
    pub process_cmdline: Option<String>,
    pub team_id: Option<String>,
    pub signing_id: Option<String>,
    pub action: String,
    pub process_tree: Vec<ProcessTreeEntry>,
}

impl ViolationEvent {
    pub fn process_name(&self) -> String {
        self.process_path
            .split('/')
            .next_back()
            .unwrap_or(&self.process_path)
            .to_string()
    }

    pub fn signing_status(&self) -> &'static str {
        if self
            .process_tree
            .first()
            .map(|e| e.is_platform_binary)
            .unwrap_or(false)
        {
            "platform"
        } else if self.team_id.is_some() {
            "signed"
        } else if self.signing_id.is_some() {
            "adhoc"
        } else {
            "unsigned"
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProcessTreeEntry {
    pub pid: u32,
    pub ppid: Option<u32>,
    pub name: String,
    pub path: String,
    pub cwd: Option<String>,
    pub cmdline: Option<String>,
    pub uid: Option<u32>,
    pub euid: Option<u32>,
    pub team_id: Option<String>,
    pub signing_id: Option<String>,
    #[serde(default)]
    pub is_platform_binary: bool,
    #[serde(default)]
    pub state: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentStatus {
    pub mode: String,
    #[serde(default)]
    pub degraded_mode: bool,
    #[serde(default)]
    pub events_pending: usize,
    #[serde(default)]
    pub connected_clients: usize,
    pub uptime_secs: u64,
    pub total_violations: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Exception {
    pub id: i64,
    pub process_path: Option<String>,
    pub signer_type: Option<String>,
    pub team_id: Option<String>,
    pub signing_id: Option<String>,
    pub file_pattern: String,
    #[serde(default)]
    pub is_glob: bool,
    pub expires_at: Option<DateTime<Utc>>,
    pub added_by: String,
    pub comment: Option<String>,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct AddExceptionParams {
    pub process_path: Option<String>,
    pub signer_type: Option<String>,
    pub team_id: Option<String>,
    pub signing_id: Option<String>,
    pub file_pattern: String,
    #[serde(default)]
    pub is_glob: bool,
    pub expires_at: Option<DateTime<Utc>>,
    pub comment: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Category {
    pub id: String,
    pub enabled: bool,
    pub patterns: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "action", rename_all = "snake_case")]
pub enum Request {
    Subscribe {
        #[serde(default)]
        filter: Option<EventFilter>,
    },
    Status,
    SetMode {
        mode: String,
    },
    AllowOnce {
        event_id: String,
    },
    AllowPermanently {
        event_id: String,
        #[serde(default)]
        expires_at: Option<DateTime<Utc>>,
        #[serde(default)]
        comment: Option<String>,
    },
    Kill {
        event_id: String,
    },
    GetViolations {
        #[serde(default)]
        limit: Option<usize>,
        #[serde(default)]
        since: Option<DateTime<Utc>>,
    },
    GetExceptions,
    #[serde(rename = "add_exception")]
    AddException(AddExceptionParams),
    RemoveException {
        id: i64,
    },
    GetCategories,
    SetCategoryEnabled {
        category_id: String,
        enabled: bool,
    },
    ResumeProcess {
        pid: u32,
    },
    Ping,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct EventFilter {
    #[serde(default)]
    pub file_patterns: Vec<String>,
    #[serde(default)]
    pub rule_ids: Vec<String>,
    #[serde(default)]
    pub denied_only: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "status", rename_all = "snake_case")]
pub enum Response {
    Success {
        #[serde(default)]
        message: String,
    },
    Error {
        message: String,
    },
    #[serde(rename = "status")]
    Status {
        mode: String,
        #[serde(default)]
        degraded_mode: bool,
        #[serde(default)]
        events_pending: usize,
        #[serde(default)]
        connected_clients: usize,
        uptime_secs: u64,
        total_violations: u64,
    },
    Violations {
        events: Vec<ViolationEvent>,
    },
    Exceptions {
        rules: Vec<Exception>,
    },
    Categories {
        categories: Vec<Category>,
    },
    Event(ViolationEvent),
    Pong,
}

fn unexpected_response() -> io::Error {
    io::Error::new(io::ErrorKind::InvalidData, "Unexpected response")
}

pub struct IpcClient {
    stream: Option<BufReader<UnixStream>>,
}

impl IpcClient {
    pub fn new() -> Self {
        Self { stream: None }
    }

    pub async fn connect(&mut self) -> io::Result<()> {
        let stream = UnixStream::connect(SOCKET_PATH).await?;
        self.stream = Some(BufReader::new(stream));
        Ok(())
    }

    pub async fn disconnect(&mut self) {
        self.stream = None;
    }

    async fn send(&mut self, request: &Request) -> io::Result<Response> {
        let stream = self
            .stream
            .as_mut()
            .ok_or_else(|| io::Error::new(io::ErrorKind::NotConnected, "Not connected"))?;

        let json = serde_json::to_string(request)?;
        stream.get_mut().write_all(json.as_bytes()).await?;
        stream.get_mut().write_all(b"\n").await?;
        stream.get_mut().flush().await?;

        let mut line = String::new();
        stream.read_line(&mut line).await?;

        serde_json::from_str(&line).map_err(|e| {
            io::Error::new(
                io::ErrorKind::InvalidData,
                format!("Parse error: {} in: {}", e, line.trim()),
            )
        })
    }

    pub async fn subscribe(&mut self, filter: Option<EventFilter>) -> io::Result<()> {
        match self.send(&Request::Subscribe { filter }).await? {
            Response::Success { .. } => Ok(()),
            Response::Error { message } => Err(io::Error::other(message)),
            _ => Err(unexpected_response()),
        }
    }

    pub async fn read_event(&mut self) -> io::Result<Option<ViolationEvent>> {
        let stream = self
            .stream
            .as_mut()
            .ok_or_else(|| io::Error::new(io::ErrorKind::NotConnected, "Not connected"))?;

        let mut line = String::new();
        if stream.read_line(&mut line).await? == 0 {
            return Ok(None);
        }

        let response: Response = serde_json::from_str(&line)
            .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e.to_string()))?;

        match response {
            Response::Event(event) => Ok(Some(event)),
            _ => Ok(None),
        }
    }

    pub async fn get_status(&mut self) -> io::Result<AgentStatus> {
        match self.send(&Request::Status).await? {
            Response::Status {
                mode,
                degraded_mode,
                events_pending,
                connected_clients,
                uptime_secs,
                total_violations,
            } => Ok(AgentStatus {
                mode,
                degraded_mode,
                events_pending,
                connected_clients,
                uptime_secs,
                total_violations,
            }),
            Response::Error { message } => Err(io::Error::other(message)),
            _ => Err(unexpected_response()),
        }
    }

    pub async fn set_mode(&mut self, mode: String) -> io::Result<()> {
        match self.send(&Request::SetMode { mode }).await? {
            Response::Success { .. } => Ok(()),
            Response::Error { message } => Err(io::Error::other(message)),
            _ => Err(unexpected_response()),
        }
    }

    pub async fn get_violations(
        &mut self,
        limit: Option<usize>,
        since: Option<DateTime<Utc>>,
    ) -> io::Result<Vec<ViolationEvent>> {
        match self.send(&Request::GetViolations { limit, since }).await? {
            Response::Violations { events } => Ok(events),
            Response::Error { message } => Err(io::Error::other(message)),
            _ => Err(unexpected_response()),
        }
    }

    pub async fn get_exceptions(&mut self) -> io::Result<Vec<Exception>> {
        match self.send(&Request::GetExceptions).await? {
            Response::Exceptions { rules } => Ok(rules),
            Response::Error { message } => Err(io::Error::other(message)),
            _ => Err(unexpected_response()),
        }
    }

    pub async fn add_exception(&mut self, params: AddExceptionParams) -> io::Result<()> {
        match self.send(&Request::AddException(params)).await? {
            Response::Success { .. } => Ok(()),
            Response::Error { message } => Err(io::Error::other(message)),
            _ => Err(unexpected_response()),
        }
    }

    pub async fn remove_exception(&mut self, id: i64) -> io::Result<()> {
        match self.send(&Request::RemoveException { id }).await? {
            Response::Success { .. } => Ok(()),
            Response::Error { message } => Err(io::Error::other(message)),
            _ => Err(unexpected_response()),
        }
    }

    pub async fn allow_once(&mut self, event_id: String) -> io::Result<()> {
        match self.send(&Request::AllowOnce { event_id }).await? {
            Response::Success { .. } => Ok(()),
            Response::Error { message } => Err(io::Error::other(message)),
            _ => Err(unexpected_response()),
        }
    }

    pub async fn allow_permanently(
        &mut self,
        event_id: String,
        expires_at: Option<DateTime<Utc>>,
        comment: Option<String>,
    ) -> io::Result<()> {
        match self
            .send(&Request::AllowPermanently {
                event_id,
                expires_at,
                comment,
            })
            .await?
        {
            Response::Success { .. } => Ok(()),
            Response::Error { message } => Err(io::Error::other(message)),
            _ => Err(unexpected_response()),
        }
    }

    pub async fn kill_process(&mut self, event_id: String) -> io::Result<()> {
        match self.send(&Request::Kill { event_id }).await? {
            Response::Success { .. } => Ok(()),
            Response::Error { message } => Err(io::Error::other(message)),
            _ => Err(unexpected_response()),
        }
    }

    pub async fn get_categories(&mut self) -> io::Result<Vec<Category>> {
        match self.send(&Request::GetCategories).await? {
            Response::Categories { categories } => Ok(categories),
            Response::Error { message } => Err(io::Error::other(message)),
            _ => Err(unexpected_response()),
        }
    }

    pub async fn set_category_enabled(
        &mut self,
        category_id: String,
        enabled: bool,
    ) -> io::Result<()> {
        match self
            .send(&Request::SetCategoryEnabled {
                category_id,
                enabled,
            })
            .await?
        {
            Response::Success { .. } => Ok(()),
            Response::Error { message } => Err(io::Error::other(message)),
            _ => Err(unexpected_response()),
        }
    }

    pub async fn resume_process(&mut self, pid: u32) -> io::Result<()> {
        match self.send(&Request::ResumeProcess { pid }).await? {
            Response::Success { .. } => Ok(()),
            Response::Error { message } => Err(io::Error::other(message)),
            _ => Err(unexpected_response()),
        }
    }
}
